cmd/internal/flags/admin.go

```diff
@@ -115,34 +113,38 @@ func (s *AdminServer) Serve() error {
 	return nil
 }
 
-func (s *AdminServer) serveWithListener(l net.Listener) {
+func (s *AdminServer) serveWithListener(l net.Listener) (err error) {
 	s.logger.Info("Mounting health check on admin server", zap.String("route", "/"))
 	s.mux.Handle("/", s.hc.Handler())
 	version.RegisterHandler(s.mux, s.logger)
 	s.registerPprofHandlers()
 	recoveryHandler := recoveryhandler.NewRecoveryHandler(s.logger, true)
 	errorLog, _ := zap.NewStdLogAt(s.logger, zapcore.ErrorLevel)
-	s.server = &http.Server{
-		Handler:           recoveryHandler(s.mux),
-		ErrorLog:          errorLog,
-		ReadHeaderTimeout: 2 * time.Second,
+	s.server, err = s.serverCfg.ToServer(context.Background(), nil, telemetry.NoopSettings().ToOtelComponent(),
+		recoveryHandler(s.mux))
+	if err != nil {
+		return err
 	}
-	if s.tlsCfg != nil {
-		s.server.TLSConfig = s.tlsCfg
+	s.server.ErrorLog = errorLog
+
+	if s.serverCfg.TLSSetting != nil {
+		tlsConfig, err := s.serverCfg.TLSSetting.LoadTLSConfig(context.Background())
+		if err != nil {
+			s.logger.Error("failed to load tls config", zap.Error(err))
+			s.hc.Set(healthcheck.Broken)
+		}
+		s.server.TLSConfig = tlsConfig
 	}
-	s.logger.Info("Starting admin HTTP server", zap.String("http-addr", s.adminHostPort))
+	s.logger.Info("Starting admin HTTP server", zap.String("http-addr", s.serverCfg.Endpoint))
 	go func() {
-		var err error
-		if s.tlsCfg != nil {
-			err = s.server.ServeTLS(l, "", "")
-		} else {
-			err = s.server.Serve(l)
-		}
+		err := s.server.Serve(l)
```

Comment:
might want to use a waitgroup here to block return at L147 until the listen goroutine has started